persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -306,101 +309,122 @@ static class PossibleExpirationNotification {
+                     } else {
+                        log.errorf("Expected %d entries to be deleted, but deleteresult contains only %d entries");
                      }
                   }
-
-                  if (list == null || !list.isEmpty()) {
-                     purgedAmount += runBatchAndNotify(list, batchDelete, unicastProcessor);
-                  }
-
-                  if (log.isTraceEnabled()) {
-                     log.tracef("Successfully purged %d rows.", purgedAmount);
-                  }
-                  conn.commit();
-                  unicastProcessor.onComplete();
+               } else {
+                  log.errorf("Expected %d entries to be deleted, but deleteresult is empty");
                }
-            } catch (SQLException e) {
-               log.failedClearingJdbcCacheStore(e);
-               try {
-                  conn.rollback();
-               } catch (SQLException ex) {
-                  log.sqlFailureTxRollback(ex);
+            }
+         } while (!entriesToBeDeleted.isEmpty());
+         if (purgedAmount > 0) {
+            log.infof("Successfully purged %d rows of table %s.", purgedAmount, getTableManager().getDataTableName());
+         }
+         unicastProcessor.onComplete();
+      } catch (SQLException e) {
+         log.failedClearingJdbcCacheStore(e);
+         unicastProcessor.onError(e);
+      }
+   }
+
+   private List<PossibleExpirationNotification> getEntriesToBeDeleted() throws SQLException {
+      int batchSize = this.configuration.maxBatchSize();
+      String sql = this.getTableManager().getSelectOnlyExpiredRowsSql();
+      try (var conn = this.connectionFactory.getConnection()) {
+         conn.setAutoCommit(false);
+         try(PreparedStatement ps = conn.prepareStatement(sql)) {
+            ps.setLong(1, this.timeService.wallClockTime());
+            try(ResultSet rs = ps.executeQuery()) {
+               List<PossibleExpirationNotification> result = new ArrayList<>();
+               while (rs.next() && (batchSize > 0)) {
+                  batchSize--;
+                  String keyStr = rs.getString(2);
+                  long expiryTime = rs.getLong(3);
+                  InputStream inputStream = rs.getBinaryStream(1);
+                  MarshalledValue value = JdbcUtil.unmarshall(inputStream, this.marshaller);
+                  result.add(new PossibleExpirationNotification(expiryTime, keyStr, value));
                }
-               unicastProcessor.onError(e);
-            } finally {
-               JdbcUtil.safeClose(rs);
-               JdbcUtil.safeClose(ps);
-               connectionFactory.releaseConnection(conn);
+               return Collections.unmodifiableList(result);
             }
-         }, "jdbcstringstore-purge");
-         return unicastProcessor;
-      });
+         } finally {
+            conn.commit();
+            conn.setAutoCommit(true);
```

Comment:
We shouldn't be always setting this to true. TBH, I don't think this code needs to mess with autoCommit at all, however if we want to to be safe, we should store the autoCommit state when setting it to false so it can be restored to what it was before instead of always being true.

Reply:
removed autocommit handling from pure reading.